### PR TITLE
Makman2/linter update

### PR DESCRIPTION
**Pull request by Makman2**

### CODE REVIEW

**Comment by Adrianzatreanu** [comment]
ack c29c09e6e

File: docs/Users/Tutorials/Linter_Bears.rst [hunk 1]
```diff
@@ -205,6 +205,10 @@ We can test our bear like this
     by their **filename** and display them based on their
     **classname**.
 
+Normally providing a severity-map is not needed, as *coala* has a default
+severity-map which recognizes many common words used for severities. Check out
+the API-documentation for keywords supported!
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/Linter_Bears.rst]: i dont think the hyphen is needed here

**Comment by Adrianzatreanu** [comment]
ack 73d25cb32156e106fae5e7db8e9ac4564e23445d

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+=========================================
+
+Often linters are no easy tools. To squeeze out the last bit of functionality
+and efficiency, ``@linter`` provides some advanced features ready for use.
+
+Supplying Configuration Files with ``generate_config``
+------------------------------------------------------
+
+Sometimes tools require a configuration file to run. ``@linter`` supports that
+easily with overriding ``generate_config()``.
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: i think ``by`` is correct here, not with

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                           "value=2 = 2")
+            return config_file
+
+        @staticmethod
+        def create_arguments(filename, file, config_file):
+            return "--use-config", config_file
+
+.. note::
+
+    By default no configuration file is generated.
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: some places its ``configuration file``, others is ``configuration-file``, and L24 its config-file. maintain consistency, i vote for the first one, with a space :)

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+parsing/processing behaviour, ``process_output`` can be overridden.
+
+::
+
+    @linter(executable='my_tool')
+    class MyBear:
+        def process_output(self, output, filename, file):
+            pass
+
+The ``output`` variable contains the string outputted from the executable.
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: i think the past tense of output is output, not outputted

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            yield Result.from_values(origin=self,
+                                     message=issue["message"],
+                                     file=filename)
+
+Additional Prerequisite Check
+-----------------------------
+
+``@linter`` supports doing an additional executable check before running the
+bear together with the normal one (checking if the executable exists). For
+example this is useful to test the existence of external modules (like Java
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: comma after example

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        for issue in json.loads(output):
+            yield Result.from_values(origin=self,
+                                     message=issue["message"],
+                                     file=filename)
+
+Additional Prerequisite Check
+-----------------------------
+
+``@linter`` supports doing an additional executable check before running the
+bear together with the normal one (checking if the executable exists). For
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: comma after bear i think

Comment (Adrianzatreanu) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: to express that the check is together with the normal one

File: coalib/bearlib/abstractions/Linter.py [hunk 7]
```diff
@@ -306,8 +306,7 @@ def process_output_corrected(self,
             Processes the executable's output as a corrected file.
```
Comment (Adrianzatreanu) [inline coalib/bearlib/abstractions/Linter.py]: commit message, line 3: where->were

**Comment by Makman2** [comment]
reack 9eda687 1708d2f

**Comment by Adrianzatreanu** [comment]
ack 624a98304dbee115c7b2118c3f8332048e2295dd

File: docs/Users/Tutorials/Linter_Bears.rst [hunk 8]
```diff
@@ -205,6 +205,10 @@ We can test our bear like this
     by their **filename** and display them based on their
     **classname**.
 
+Normally providing a severity-map is not needed, as *coala* has a default
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/Linter_Bears.rst]: i think comma is needed after Normally

**Comment by Adrianzatreanu** [comment]
ack 93020cf84ffd698f216e28976976a9f3cb0d4cf7

**Comment by Adrianzatreanu** [comment]
unack d5d643036205a94b452051349a73d7441baf2e27

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+                           "value=2 = 2")
+            return config_file
+
+        @staticmethod
+        def create_arguments(filename, file, config_file):
+            return "--use-config", config_file
+
+.. note::
+
+    By default no configuration file is generated.
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: comma after default

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+Custom Processing Functions with ``process_output``
+---------------------------------------------------
+
+Inside ``@linter`` only a few fixed outputs are supported. And they can't be
+combined for different output streams. Specifying an own output
+parsing/processing behaviour, ``process_output`` can be overridden.
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: im sorry but this last sentence doesnt really make sense to me, maybe rephrase it or explain it as i dont really get its point?

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    class MyBear:
+        def process_output(self, output, filename, file):
+            pass
+
+The ``output`` variable contains the string output from the executable.
+Depending on how you use the ``use_stdout`` and ``use_stderr`` parameters from
+``@linter``, ``output`` can contain either a tuple or a plain string: If
+``use_stdout`` and ``use_stderr`` are both ``True``, a tuple is placed with
+``(stdout, stderr)``. If only one of them is ``True``, a string is passed
+(containing the regarding activated output stream).
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: maybe ``containing the output stream(s) chosen`` sounds better :D ?

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+            yield Result.from_values(origin=self,
+                                     message=issue["message"],
+                                     file=filename)
+
+Additional Prerequisite Check
+-----------------------------
+
+``@linter`` supports doing an additional executable check before running the
+bear, together with the normal one (checking if the executable exists). For
+example, this is useful to test the existence of external modules (like Java
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: to test for the

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+::
+
+    @linter(executable='...'
+            prerequisite_check_command=('python3', '-c', 'import my_module'))
+    class MyBear:
+        pass
+
+If the default error message does not suit you, you can also supply
+``prerequisite_check_fail_message`` together with
+``prerequisite_check_command``.
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: mention that if you do not provide ``check_command``, the ``check_fail_message`` wont work :D ?

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst (hunk 10 above)
Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: yeah this sentence went kind of bad :3

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst (hunk 13 above)
Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: hm not sure, imo it does not make sense after this sentence to supply one without `check_command`. And if someone tries he will get an error anyway :3

**Comment by Makman2** [comment]
reack e123a36

**Comment by Adrianzatreanu** [comment]
rebase needed, and somehow we need that coverage down too

**Comment by Adrianzatreanu** [comment]
ack c54a1ba07bd00abc25d3c8ce5fe71fa3fd266d4b

**Comment by Makman2** [comment]
reack
740b4a4 Self_Correcting_Bears.rst: Delete
1c4722d Linter_Bears.rst: Refer to API for severities
087662e linter: Provide a better skip-reason in cwd-test
0fd85c5 linter: Use adapter function for each exe input
df5fd20 Linter: Make class fields "private"

**Comment by Makman2** [comment]
reack
fcf0f47 docs: Add advanced feature reference for linter
94f23e4 Self_Correcting_Bears.rst: Delete
67c5dea Linter_Bears.rst: Refer to API for severities
6c5fb29 linter: Provide a better skip-reason in cwd-test
3fc0312 linter: Use adapter function for each exe input
db3f9ac Linter: Make class fields "private"

**Comment by Makman2** [comment]
unack fcf0f47 docs: Add advanced feature reference for linter
not yet reviewed

**Comment by Makman2** [comment]
reack
e81114b linter: Provide a better skip-reason in cwd-test 		
ef6658c Linter_Bears.rst: Refer to API for severities
4a18ea8 Self_Correcting_Bears.rst: Delete

**Comment by Makman2** [comment]
reack
b609f90 linter: Provide a better skip-reason in cwd-test 		
dc93fd3 Linter_Bears.rst: Refer to API for severities
c423ab7 Self_Correcting_Bears.rst: Delete

**Comment by Adrianzatreanu** [comment]
ack 267f1e74d49b34794a05918298fb7d391f9fa8b1

**Comment by Adrianzatreanu** [comment]
ack 2cf4052c4916beffaf71a08bb981e498d4ab3fa0

**Comment by Adrianzatreanu** [comment]
ack 914640dd5f91cef94b283130fe028d118c3ca887

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+.. note::
+
+    By default, no configuration file is generated.
+
+Custom Processing Functions with ``process_output``
+---------------------------------------------------
+
+Inside ``@linter`` only a few output formats are supported. And they can't be
+combined for different output streams. To specify an own output
+parsing/processing behaviour, ``process_output`` can be overridden.
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: i still dont think this paragraph flows nicely. let me try, maybe you like some parts and you mix them in:

Inside @linter there are only a few output formats which are supported. To specify an own
output parsing or pcesossing behaviour, simply override ``process_output``.
However, the output formats can't be combined for different output streams.

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+Depending on how you use the ``use_stdout`` and ``use_stderr`` parameters from
+``@linter``, ``output`` can contain either a tuple or a plain string: If
+``use_stdout`` and ``use_stderr`` are both ``True``, a tuple is placed with
+``(stdout, stderr)``. If only one of them is ``True``, a string is passed
+(containing the output stream chosen).
+
+Inside ``process_output`` you need to yield results according to the executable
+output. It is also possible to combine the built-in capabilities. There are
+several functions accessible with the naming scheme
+``process_output_<output-format>``.
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: should end with ``:`` ?

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+
+- ``process_output_regex``: Extracts results using a regex.
+- ``process_output_corrected``: Extracts results (with patches) by using a
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: the by is redundant

**Comment by Adrianzatreanu** [comment]
just small changes on the last one, rest looks good :D

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst (hunk 14 above)
Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: Like that, though the last sentence is actually wrong, this is one of the advantages that you can now combine different output formats for different output streams :)

File: tests/output/JSONEncoderTest.py [hunk 17]
```diff
@@ -75,6 +76,11 @@ def test_time(self):
         self.assertEquals('"' + tf.isoformat() + '"',
                           json.dumps(tf, **self.kw))
 
+    def test_re_object(self):
+        uut = re.compile('x')
+        self.assertEqual('"' + uut.pattern + '"',
+                         json.dumps(uut, **self.kw))
```
Comment (sils1297) [inline tests/output/JSONEncoderTest.py]: yeeeees our jsonencoder is a wonder thing meant to deal with about anything you throw at it :)

File: coalib/bearlib/abstractions/Linter.py [hunk 18]
```diff
@@ -363,8 +358,7 @@ def process_output_regex(
             Processes the executable's output using a regex.
```
Comment (sils1297) [inline coalib/bearlib/abstractions/Linter.py]: you should mention in both places that this may be called twice with stderr *and* stdout if specified so

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala/pull/2588/commits/914640dd5f91cef94b283130fe028d118c3ca887

File: tests/bearlib/abstractions/LinterTest.py [hunk 19]
```diff
@@ -591,7 +591,8 @@ def test_repr(self):
             re.escape(repr(sys.executable)) + "\\) at 0x[a-fA-F0-9]+>")
 
     @skipIf(platform.system() == "Windows",
-            "Nobody can sanely test things on windows")
+            "`pwd` does not exist in Windows-cmd and `cd` is a built-in "
+            "command which fails the executable-existence check from @linter.")
```
Comment (sils1297) [inline tests/bearlib/abstractions/LinterTest.py]: really? The old one was so much better!

File: docs/Users/Tutorials/Linter_Bears.rst [hunk 20]
```diff
@@ -205,6 +205,10 @@ We can test our bear like this
     by their **filename** and display them based on their
     **classname**.
 
+Normally, providing a severity-map is not needed, as *coala* has a default
+severity-map which recognizes many common words used for severities. Check out
+the API documentation for keywords supported!
```
Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears.rst]: please link to it (hint: are you even able to find it? Our API docs suck. Don't expect the user to crawl them.)

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala/pull/2588/commits/dc93fd33c1a32ac1c315a44053ea7f3826ec4662

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst [hunk 21]
```diff
@@ -0,0 +1,134 @@
+Linter Bears - Advanced Feature Reference
```
Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: can you place links to the in the other tutorial at meaningful places?

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+    class MyBear:
+        @staticmethod
+        def generate_config(filename, file):
+            config_file = ("value1 = 1\n"
+                           "value=2 = 2")
+            return config_file
+
+The string returned from this method is written into a temporary file before
+invoking ``create_arguments()``. If you return ``None``, no configuration file
+is generated.
```
Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: please mention that you can make this a non static method if you whish to use bear logging

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+::
+
+    @linter(executable='my_tool')
+    class MyBear:
+        def process_output(self, output, filename, file):
+            pass
+
+The ``output`` variable contains the string output from the executable.
+Depending on how you use the ``use_stdout`` and ``use_stderr`` parameters from
+``@linter``, ``output`` can contain either a tuple or a plain string: If
```
Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: not true anymore, huh?

Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: oh sorry, right

File: coalib/bearlib/abstractions/Linter.py [hunk 24]
```diff
@@ -444,9 +434,17 @@ def process_output_regex(
                                 "result_message")
                     if key in options}
 
-                process_output = partialmethod(
+                _processing_function = partialmethod(
                     process_output_regex, **_process_output_args)
 
+            def process_output(self, output, filename, file):
```
Comment (sils1297) [inline coalib/bearlib/abstractions/Linter.py]: please make a good documentation comment, people will look here when overriding this function

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala/pull/2588/commits/debd65060483340a4e2c7ce254caeff261f6dff3

File: tests/bearlib/abstractions/LinterTest.py (hunk 19 above)
Comment (Makman2) [inline tests/bearlib/abstractions/LinterTest.py]: nope :3

File: coalib/bearlib/abstractions/Linter.py (hunk 18 above)
Comment (Makman2) [inline coalib/bearlib/abstractions/Linter.py]: doesn't matter from user perspective imo^^

File: docs/Users/Tutorials/Linter_Bears_Advanced.rst (hunk 22 above)
Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears_Advanced.rst]: maybe I should put this into the basic tutorial as a note, I believe it fits better there^^

**Comment by Makman2** [comment]
reack 5871f6d 67ed452 cf33fcc d5d5bb3

File: docs/Users/Tutorials/Linter_Bears.rst [hunk 25]
```diff
@@ -387,4 +387,4 @@ Where to Find More...
 ---------------------
 
 If you need more information about the ``@linter`` decorator, refer to the API
-documentation. It's more versatile than described in this little tutorial.
+documentation.
```
Comment (SanketDG) [inline docs/Users/Tutorials/Linter_Bears.rst]: link to the API docs

**Comment by SanketDG** [comment]
ack 1fb4132 2b8fdbd

**Comment by abhsag24** [comment]
unack 6b225e4 error in commti message :P

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala/pull/2588/commits/212d084173eda88651702965348564a3791f8e1d

File: coalib/bearlib/abstractions/Linter.py (hunk 18 above)
Comment (sils1297) [inline coalib/bearlib/abstractions/Linter.py]: it does if you override this function!

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala/pull/2588/commits/9fc2eb64d218b26a5bb8168d2ccdad319e114380

File: coalib/bearlib/abstractions/Linter.py [hunk 26]
```diff
@@ -444,9 +434,34 @@ def process_output_regex(
...
+                    ``use_stderr`` parameters of ``@linter``. If only one of
+                    these arguments is ``True``, a string is placed (containing
+                    the selected output stream). If both are ``True``, a tuple
+                    is placed with ``(stdout, stderr)``.
+                :param filename:
+                    The name of the file currently processed.
+                :param file:
+                    The contents of the file (line-splitted).
+                """
+
```
Comment (sils1297) [inline coalib/bearlib/abstractions/Linter.py]: I'd personally remove this newline but that's a taste thing, keep if if you want

File: coalib/bearlib/abstractions/Linter.py (hunk 18 above)
Comment (Makman2) [inline coalib/bearlib/abstractions/Linter.py]: why should someone override it?
if someone wants custom parsing behaviour, there's no need for doing that. Use your own `process_output` which allows way more control^^

File: docs/Users/Tutorials/Linter_Bears.rst (hunk 25 above)
Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears.rst]: I don't know how to do that :3

File: docs/Users/Tutorials/Linter_Bears.rst (hunk 20 above)
Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears.rst]: how do I link to the api docs? especially to a specific class?
btw the linter API docs are still not working, so for now it makes no difference, I think we should file an issue for that^^

File: docs/Users/Tutorials/Linter_Bears.rst (hunk 25 above)
Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears.rst]: as the linter API is still not showing up, mind filing an issue for that after merge? (or even now)

File: docs/Users/Tutorials/Linter_Bears.rst (hunk 20 above)
Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears.rst]: there's an issue placing this link at the bottom of the page (the API is already mentioned there). I would open another one on merge^^

**Comment by Makman2** [comment]
reack
546faea Linter_Bears.rst: Use title case
52c05d3 Linter_Bears.rst: Wrap text
0a3c002 Self_Correcting_Bears.rst: Delete
68742fa linter: Provide a better skip-reason in cwd-test
41e3517 Linter: Make class fields "protected"
25e5ce2 JSONEncoder: Add test for SRE_pattern type

**Comment by sils1297** [comment]
ouh @Makman2 IIUC you'll want to remove all the doc changes as @Adrianzatreanu is moving them to the doc repo

**Comment by sils1297** [comment]
or rather port them, also you've got conflicts. I know it's mean but I'm marking this WIP now :P

**Comment by Makman2** [comment]
documentation commits now at https://github.com/coala-analyzer/documentation/pull/68

**Comment by Makman2** [comment]
reack 3c97b70 0268dec 46de358

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2588/commits/45bd134829afcc88ea8952bdb7c94110872be832

**Comment by sils1297** [comment]
@rultor merge
